### ISSUE TITLE
[BUG/Help] 32K model cannot run

### ISSUE THREAD

**Issue by wangjvjie** [THUDM/ChatGLM2-6B#421]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

我使用web_demo加载32k模型，但是没办法正常运行。报错RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1! (when checking argument for argument tensors in method wrapper_CUDA_cat)
Traceback:
File "/home/h3c/anaconda3/envs/pytorch/lib/python3.9/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 552, in _run_script
    exec(code, module.__dict__)
File "/data/pythonProject/ChatGLM2-6B/web_demo2.py", line 62, in <module>
    for response, history, past_key_values in model.stream_chat(tokenizer, prompt_text, history,
File "/home/h3c/anaconda3/envs/pytorch/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 35, in generator_context
    response = gen.send(None)
File "/home/h3c/.cache/huggingface/modules/transformers_modules/chatglm2-6b-32k/modeling_chatglm.py", line 1072, in stream_chat
    for outputs in self.stream_generate(**inputs, past_key_values=past_key_values,
File "/home/h3c/anaconda3/envs/pytorch/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 35, in generator_context
    response = gen.send(None)
File "/home/h3c/.cache/huggingface/modules/transformers_modules/chatglm2-6b-32k/modeling_chatglm.py", line 1157, in stream_generate
    outputs = self(
File "/home/h3c/anaconda3/envs/pytorch/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
File "/home/h3c/anaconda3/envs/pytorch/lib/python3.9/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
File "/home/h3c/.cache/huggingface/modules/transformers_modules/chatglm2-6b-32k/modeling_chatglm.py", line 946, in forward
    transformer_outputs = self.transformer(
File "/home/h3c/anaconda3/envs/pytorch/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
File "/home/h3c/.cache/huggingface/modules/transformers_modules/chatglm2-6b-32k/modeling_chatglm.py", line 836, in forward
    hidden_states, presents, all_hidden_states, all_self_attentions = self.encoder(
File "/home/h3c/anaconda3/envs/pytorch/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
File "/home/h3c/.cache/huggingface/modules/transformers_modules/chatglm2-6b-32k/modeling_chatglm.py", line 655, in forward
    presents = torch.cat((presents, kv_cache), dim=0)

### Steps To Reproduce

我按照Readme中的操作，只是加载32k模型就报错，但是加载原模型没问题：
def get_model():
    tokenizer = AutoTokenizer.from_pretrained("chatglm2-6b-32k", trust_remote_code=True)
    # model = AutoModel.from_pretrained("chatglm2-6b-32k", trust_remote_code=True).cuda()
    # 多显卡支持，使用下面两行代替上面一行，将num_gpus改为你实际的显卡数量
    from utils import load_model_on_gpus
    model = load_model_on_gpus("chatglm2-6b-32k", num_gpus=4)
    model = model.eval()
    return tokenizer, model

### Environment

```markdown
- OS:
- Python:
- Transformers:
- PyTorch:
- CUDA Support (`python -c "import torch; print(torch.cuda.is_available())"`) :
```

**Comment by mayi140611** [comment]
File "/home/h3c/.cache/huggingface/modules/transformers_modules/chatglm2-6b-32k/modeling_chatglm.py", line 655, in forward
presents = torch.cat((presents, kv_cache), dim=0)
根据报错信息，在执行 torch.cat操作前 需要保证presents, kv_cache)这两个变量在同一个device上，按照这个思路改一下代码吧

**Comment by 1250681923** [comment]
解决了吗？碰见了同样的错误，烦请指点

**Comment by xtuhcy** [comment]
同问，解决了吗？

**Comment by mayi140611** [comment]
> File "/home/h3c/.cache/huggingface/modules/transformers_modules/chatglm2-6b-32k/modeling_chatglm.py", line 655, in forward presents = torch.cat((presents, kv_cache), dim=0) 根据报错信息，在执行 torch.cat操作前 需要保证presents, kv_cache)这两个变量在同一个device上，按照这个思路改一下代码吧

测试了一下，以下代码可以通过:
```
hidden_states, kv_cache = layer_ret
if use_cache:
    # token by token decoding, use tuple format
    if kv_caches[0] is not None:
        presents = presents + (kv_cache,)
    # prefilling in decoding, use tensor format to save cuda memory
    else:
        if len(presents) == 0:
            presents = kv_cache
        else:
            presents_device = presents.device
            kv_cache_device = kv_cache.device
            if kv_cache_device != presents_device:
                kv_cache = kv_cache.to(presents_device)
            if isinstance(presents, tuple):
                presents = tuple(tensor.to(presents_device) for tensor in presents)
            else:
                presents = presents.to(presents_device)
            presents = torch.cat((presents, kv_cache), dim=0)
```
